synapse/handlers/room_member.py

```diff
@@ -463,10 +447,21 @@ def _update_membership(
                 if requester.is_guest:
                     content["kind"] = "guest"
 
-                ret = yield self._remote_join(
+                remote_join_response = yield self._remote_join(
                     requester, remote_room_hosts, room_id, target, content
                 )
-                return ret
+
+                # Copy over user state if this is a join on an remote upgraded room
+                yield self.copy_user_state_on_room_upgrade(
+                    room_id, requester.user.to_string()
+                )
+
+                return remote_join_response
+
+            # Copy over user state if this is a join on an local upgraded room
+            yield self.copy_user_state_on_room_upgrade(
+                room_id, requester.user.to_string()
+            )
```

Comment:
We only want to do this after join, so I think it should be back in `_local_membership_update`

Reply:
I've moved it to `_local_membership_update`. Do you think it should be inside `_user_joined_room` as well? My only concern is I'm not entirely sure where that method runs.